docs/messaging.md

```diff
@@ -0,0 +1,126 @@
+  successful, failed, etc.
+  See [ProtocolReply.Status](/atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/ProtocolReply.java)
+  for more.
+
+### Messaging exceptions
+
+When sending messages between nodes, there's a shared set of exceptions that may be returned by the
+messaging service on top of whatever subject specific exceptions your own code may return.
+
+The recommendation is to encode business logic failures as successful protocol messages, and handle
+their own semantics near the calling site and **not** in the messaging layer.
+
+As such, exceptions returned by the messaging layer typically will indicate a real issue, and you
+should handle them whenever sending any requests through the messaging layer:
+
+- `java.lang.IllegalStateException` - indicates the underlying service cannot send the given
+  request (e.g. the service is not running)
+- `java.net.ConnectException` - indicates the recipient is unreachable; typically this is temporary,
+  and should be logged as a warning and retried.
+- `java.util.concurrent.TimeoutException` - indicates no response came back after the default
+  timeout.
+- `io.atomix.cluster.messaging.MessagingException.ProtocolException` - indicates the recipient
+  failed to parse the incoming message. This is typically not retry-able, as the encoding logic is
+  independent of the input, so if two nodes cannot encode/decode in a format they both understand,
+  there's no point retrying.
+- `io.atomix.cluster.messaging.MessagingException.NoRemoteHandler` - indicates the recipient
+  received the message, but has no handler registered for it. This can be transient - for example,
+  if the recipient is shutting down, starting up, or in a state transition such that it is
+  recreating its handlers. In such cases, this can be retried. However, if this is consistently
+  happening, then this indicates a real error.
+- `io.atomix.cluster.messaging.MessagingException.RemoteHandlerFailure` - indicates the recipient
+  parsed and expected the message, but it failed unexpectedly to process it. Whether this is
+  retry-able or not depends on the business logic around this subject, and is up to the caller to
+  decide. Best practice, however, is to handle expected business errors as successful protocol
+  replies, such that this error is returned _only_ for unexpected exceptions, which would normally
+  not be retry-able.
+
+On top of these, you also have an additional exception when using
+the [ClusterCommunicationService](/atomix/cluster/src/main/java/io/atomix/cluster/messaging/ClusterCommunicationService.java),
+which is used specifically to communicate between members of the same cluster via their IDs.
+
+- `io.atomix.cluster.messaging.MessagingException.NoSuchMemberException` - this can occur in two
+  ways. This can be returned immediately at the calling site (so on the client side) if the there is
+  known member with the given ID in the cluster (obtained via the membership protocol). It can also
```

Comment:
❓ I think you mean `no known member` or?

Reply:
:see_no_evil: Yes, no known member